### PR TITLE
Always try to append a slash

### PR DESCRIPTION
**Pull request by Xiphoseer**
This pull-request changes the behavior of the redirect resolution slightly.

1. As explained in detail in #31, the django `APPEND_SLASH` setting that defaults to `True` is used to indicate that the system should treat URLs with and without slash the same way. In the past, djangocms-redirect only tried a lookup with an added slash when that setting was disabled, but that means that the default case breaks down for sites that use i18n_patterns (see #31)
2. A `clean` method was added to the model, that enforces that `old_url` starts with a slash and, if `APPEND_SLASH` is set, also that it must end with a slash. This is consistent with the visible behavior of the `PageSmartLinkWidget`, which shows urls without a slash at the end but produces URLs with a slash for the instance.
3. I replaced the pattern of `get_full_path().partition(?)` with individual calls to `request.META['QUERY_STRING']` and `request.path`, and inverted the urlencoding logic, because the unquoted variant of a partitioned full_path is always equal to `request.path`.

(3.) broke a test, which I "fixed". The actual reason the test broke is not that the encoding logic would not work for all cases anymore, but rather that the cache invalidation logic only triggers for the exact value in the DB and so misses either the quoted or unquoted path. Because I switched to using the actual `request.path` as the cache key, instead of the quoted variant, the cache wasn't invalidated when a quoted `old_path` was stored.

Because the cache can be filled by either quoted or unquoted database entries, this is buggy either way and may be resolved by unquoting (with my patch applied) at cache invalidation. You can also get there by quoting there without my patch applied, but that runs the risk of douple-quoting if you don't also invalidate the exact value.

### CODE REVIEW

File: djangocms_redirect/middleware.py [hunk 1]
```diff
@@ -38,24 +38,31 @@ def do_redirect(self, request, response=None):
...
+            req_path_slash_quoted = escape_uri_path(req_path_slash)
+            if req_path_slash_quoted != req_path_slash:
+                possible_paths.append(req_path_slash_quoted)
+
         current_site = get_current_site(request)
         r = None
-        key = get_key_from_path_and_site(full_path_quoted, settings.SITE_ID)
+        key = get_key_from_path_and_site(req_path, settings.SITE_ID)
         cached_redirect = cache.get(key)
+        if req_path == '/path (escaped)/' and cached_redirect:
```
Comment (yakky) [inline djangocms_redirect/middleware.py:64]: why is this needed?

File: tests/test_middleware.py [hunk 2]
```diff
@@ -219,7 +219,7 @@ def test_redirect_no_append_slash(self):
     def test_redirect_no_append_slash_quoted(self):
         pages = self.get_pages()
 
-        original_path = '/path%20(escaped)/'
+        original_path = '/path (escaped)/'
```
Comment (yakky) [inline tests/test_middleware.py:222]: according to my tests we only need to adapt the number of queries run for this case, I'd like to test the quoted version too

**Review by yakky — APPROVED** [review]
@Xiphoseer first of all thanks for the very detailed explanation and the effort in handling this issue

Overall your PR looks :+1: 

I only added minor comments

I want to resolve the issues with travis before merging this

I will also merge #38 before this to avoid python 2 which are not relevant anymore

File: djangocms_redirect/middleware.py (hunk 1 above)
Comment (Xiphoseer) [reply djangocms_redirect/middleware.py:64]: It's not, thank you for catching that. It was a holdover from me trying to debug the test that failed.

File: tests/test_middleware.py (hunk 2 above)
Comment (Xiphoseer) [reply tests/test_middleware.py:222]: I duplicated the test to try both cases. I agree that they are different because the contents of the DB are different. The matching succeeds because `escape_uri_path` doesn't escape the parentheses, and the quoted case is explicitly added to the paths that are looked up.

I still think that the support for quoted strings in the DB is not obvious and maybe only in there because `get_full_path` was used previously, which returns quoted strings. It's easy to think that putting a string like `/path%20%28escaped%29/` in the DB would work, but it did and still does not, while `/path%20(escaped)/` did and does. Ideally, that topic would get some sort of mention in the docs.

**Comment by Xiphoseer** [comment]
Hey, thanks for the feedback. Are the "issues with travis" something about the code in this PR or some other problem related to the setup of this repo?

As this tests changes the way `APPEND_SLASH` is used, I'm wondering whether there are some tests, that are currenly only run with `APPEND_SLASH=False` that should really be run for both cases. From my point of view, the tests that used to be about special-casing the `False` case are now regression tests for that case, but because we can't rely on the `CommonMiddleware` to always append the slash in the `True` case (because of the locale prefix issue) we really want to have the same behavior independent of the setting and make sure that's actually the case.

File: tests/test_middleware.py (hunk 2 above)
Comment (yakky) [reply tests/test_middleware.py:222]: Could you suggest some phrasing to clarify this? You look much better than me at explaining things :joy:

**Comment by yakky** [comment]
@Xiphoseer thanks for the updates, I will go through them soon

Travis basically killed our organization in their migration to paid-only version, so we first have to migrate this repo to github actions, it has nothing to do with your PR.

#38 is required to pass tests in the CI because this PR introduces some python 3 only code, which is perfectly fine. Because of the issue with travis, I will close #38 anyway and I will create a PR that brings GA and python 3 only support

If you created this PR authorizing the mantainer edits I will take care of rebasing and fixing conflicts, because there will be many :)

IIRC all cases are tested with both values of `APPEND_SLASH` and there should be no special case handling

**Comment by Xiphoseer** [comment]
Hmm, I'm not sure whether maintainer edits are enabled, because I just pushed develop to my fork and created the PR from there. I don't recall seeing or ticking such a box for this PR.

I'll try to rebase this in a timely manner whenever necessary if it turns out that GH doesn't let you do that. I'm also fine with resubmitting the PR, in case that becomes necessary.

**Comment by yakky** [comment]
@Xiphoseer awesome, thanks

I'll merge #40 shortly, and you will be able to rebase on develop

**Comment by Xiphoseer** [comment]
The settings is visible in the desktop interface and active for this PR:
image

**Comment by yakky** [comment]
@Xiphoseer rebase seem correct
If you have some time a test for the clean method would be great!

**Comment by Xiphoseer** [comment]
Yeah, I'll try to find time to look into that tomorrow / later today, it's close to 1 AM here in Europe right now.

**Comment by yakky** [comment]
@Xiphoseer no hurry :+1: 
I'm in CET too :)

**Comment by Xiphoseer** [comment]
Ok, I've added tests for every possible control-flow path of the `clean` method by creating `RedirectForm`s and `save()`ing them.

**Review by yakky — APPROVED** [review]

**Comment by yakky** [comment]
@Xiphoseer this is just awesome :bow:
